core/src/net/tcp_stream.cpp

```diff
@@ -0,0 +1,266 @@
+    {
+        CUBOS_ERROR("Invalid socket");
+        return 0;
+    }
+
+    size_t offset = 0;
+
+    // if we have a peeked character, read it first.
+    if (mPeeked)
+    {
+        static_cast<char*>(buffer)[0] = mPeekBuffer;
+        mPeeked = false;
+        offset = 1;
+    }
+
+    if (offset < size)
+    {
+
+#ifdef _WIN32
+        auto bytesRead = ::recv(mSock, static_cast<char*>(buffer) + offset, static_cast<int>(size - offset), 0);
+#else
+        auto bytesRead = ::recv(mSock, static_cast<char*>(buffer) + offset, size - offset, 0);
+#endif
+        if (bytesRead < 0)
+        {
+            CUBOS_ERROR("Failed to read from TCP stream: {}", logSystemError());
+            return 0;
+        }
+
+        if (bytesRead == 0)
+        {
+            mEof = true;
+        }
+
+        auto bytesReadCasted = static_cast<std::size_t>(bytesRead + offset);
+        mPos += bytesReadCasted;
+        return bytesReadCasted;
+    }
+
+    mPos += offset;
+    return offset;
+}
+
+bool TcpStream::readExact(void* data, std::size_t size)
+{
+    char* dataPtr = static_cast<char*>(data);
+    std::size_t totalBytesRead = 0;
+
+    while (totalBytesRead < size)
+    {
+        auto bytesRead = read(dataPtr + totalBytesRead, size - totalBytesRead);
+        if (bytesRead <= 0)
+        {
+            return false;
+        }
+
+        totalBytesRead += bytesRead;
+    }
+
+    return true;
+}
+
+std::size_t TcpStream::write(const void* buffer, std::size_t size)
+{
+    if (mSock == InnerInvalidSocket)
+    {
+        CUBOS_ERROR("Invalid socket");
+        return 0;
+    }
+
+#ifdef _WIN32
+    auto bytesWritten = ::send(mSock, static_cast<const char*>(buffer), static_cast<int>(size), 0);
+#else
+    auto bytesWritten = ::send(mSock, buffer, size, 0);
+#endif
+    if (bytesWritten < 0)
+    {
+        CUBOS_ERROR("Failed to write TCP buffer: {}", logSystemError());
+        return 0;
+    }
+
+    auto bytesWrittenCasted = static_cast<std::size_t>(bytesWritten);
+    mPos += bytesWrittenCasted;
+    return bytesWrittenCasted;
+}
+
+bool TcpStream::writeExact(const void* data, std::size_t size)
+{
+    const char* dataPtr = static_cast<const char*>(data);
+    std::size_t totalBytesRead = 0;
+
+    while (totalBytesRead < size)
+    {
+        auto bytesWritten = write(dataPtr + totalBytesRead, size - totalBytesRead);
+        if (bytesWritten <= 0)
+        {
+            return false;
+        }
+
+        totalBytesRead += bytesWritten;
+    }
+
+    return true;
+}
+
+std::size_t TcpStream::tell() const
+{
+    CUBOS_FAIL("Not implemented");
+}
+
+char TcpStream::peek()
+{
+    if (mSock == InnerInvalidSocket)
+    {
+        CUBOS_ERROR("Invalid socket");
+        return '\0';
+    }
+
+    if (mPeeked)
+    {
+        return mPeekBuffer;
+    }
+
+    char c;
+
+#ifdef _WIN32
+    auto bytesRead = ::recv(mSock, &c, 1, MSG_PEEK);
+#else
+    auto bytesRead = ::recv(mSock, static_cast<void*>(&c), 1, MSG_PEEK);
```

Comment:
```suggestion
    auto bytesRead = ::recv(mSock, &mPeekBuffer, 1, MSG_PEEK);
#else
    auto bytesRead = ::recv(mSock, static_cast<void*>(&mPeekBuffer), 1, MSG_PEEK);
```
The `c` variable isn't needed here right?